File: Clients-WorkersNode/Workers/worker_node.rb
require 'socket'
require 'thread'

class StreamSocket
  def initialize(socket)
    @socket = socket
  end

  def self.connect(ip, port)
    new(TCPSocket.new(ip, port))
  rescue StandardError => e
    puts "Connection error: #{e.message}"
    exit(1)
  end

  def receive
    @socket.gets
  rescue StandardError => e
    puts "Receive error: #{e.message}"
    exit(1)
  end

  def send(message)
    @socket.puts(message)
  rescue StandardError => e
    puts "Send error: #{e.message}"
    exit(1)
  end

  def close
    @socket.close
  end
end

class WorkerBeat
  def initialize(server)
    @server = server
  end

  def start
    Thread.new do
      loop do
        begin
          sb = @server.receive
          puts sb
          puts 'Sending the heartbeat to the server....'
          @server.send("HeartBeat from the worker node!!\n")
          sleep 1
        rescue StandardError => e
          puts "Heartbeat error: #{e.message}"
          break
        end
      end
      puts 'WorkerBeat finalizado.'
    end
  end
end

class WorkerNode
  def self.count_a(s)
    s.count('a')
  end

  def self.run
    begin
      port = 8084
      ip = '10.10.0.231'
      server = StreamSocket.connect(ip, port)
      puts 'Connected to the leader node...'
      server_beat = StreamSocket.connect(ip, port)
      puts 'Creating the heartbeat channel to the server...'
      puts 'Starting the heartbeats...'

      WorkerBeat.new(server_beat).start

      loop do
        puts 'Waiting for data...'
        msg = server.receive
        puts 'Received Data:'
        puts msg
        puts ''

        ans = ''
        msg.each_line do |line|
          next if line.strip.empty?

          puts "line: #{line}"
          line.split.each do |word|
            cnt = count_a(word)
            ans += "#{cnt}\n"
          end
        end

        puts 'Processed Data:'
        puts ans
        server.send(ans)
      end
    rescue StandardError => e
      puts "Exception: #{e.message}"
    end
  end
end

WorkerNode.run

File: Clients-WorkersNode/Workers/WorkerNode.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package heart_beat;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author jhozzel
 */
public class WorkerNode {
    
    public static int f(String s) { // count the numbers of 'a'
        int cnt = 0;
        for (char c : s.toCharArray()) {
            if (c == 'a') cnt++;
        }
        return cnt;
    }
    
    public static void main(String[] args) {
        try {
            int PUERTO = 8084;
            String ip = "localhost";
            StreamSocket server = new StreamSocket(new Socket(ip, PUERTO));
            System.out.println("Connected to the leader node...");
            StreamSocket serverBeat = new StreamSocket(new Socket(ip, PUERTO));
            System.out.println("Creating the heartbeat channel to the server...");
            System.out.println("Starting the heartbeats...");
            
            new WorkerBeat(serverBeat).start();
            while (true) {
                System.out.println("Waiting for data...");
                StringBuilder mssg = server.receive();
                System.out.println("Received Data: ");
                
                
                System.out.println(mssg.toString());
                System.out.println("");
                
                // Process the data
                Scanner sc = new Scanner(mssg.toString());
                String line = sc.nextLine();
                int n = Integer.parseInt(line);
                ArrayList<String> a = new ArrayList<>();
                for (int i = 0; i < n; i++) {
                    line = sc.nextLine();
                    // Dividir la línea en palabras usando split()
                    String[] palabras = line.split("\\s+"); // Esto divide por espacios en blanco

                    // Agregar cada palabra al ArrayList
                    for (String palabra : palabras) {
                        a.add(palabra);
                    }
                }
                n = a.size();

                line = sc.nextLine();
                int k = Integer.parseInt(line);
                ArrayList<String> b = new ArrayList<>();
                for (int i = 0; i < k; i++) {
                    line = sc.nextLine();
                    
                    b.add(line);
                }
                int[] freq = new int [k];
                for (int i = 0; i < k; i++) {
                    freq[i] = 0;
                }
                for (int i = 0; i < n; i++) {
                    for (int j = 0; j < k; j++) {
                        if (b.get(j).equals(a.get(i))) {
                            freq[j]++;
                        }
                    }
                }
                
                StringBuilder ans = new StringBuilder();
                for (int i = 0; i < k; i++) {
                    line = String.valueOf(freq[i]);
                    ans.append(line).append('\n');
                }
                
                System.out.println("Proccesed Data: ");
                System.out.println(ans.toString());
                server.send(ans);
            }
        } catch (IOException ex) {
            Logger.getLogger(Cliente.class.getName()).log(Level.SEVERE, null, ex);
        }
    }
}
